app/services/exchanger_service.rb

```diff
@@ -0,0 +1,29 @@
+class ExchangerService
+  def initialize(params)
+    @params = params
+  end
+
+  def call(currency = CurrencyService.new(params[:date]))
+    return unless params_present?
```

Comment:
Should we not return a error if all the required fields are not present? At the moment this will cause an error because in the controller you set the service to a variable, then if it returns nil as here, it'll go to failed. The problem is that you're calling response.errors in the failed status as well, so it'll cause a 500 error.